crates/fluvio/src/producer/dispatcher.rs

```diff
@@ -0,0 +1,219 @@
+use std::pin::Pin;
+use std::sync::Arc;
+use std::collections::BTreeMap;
+use futures_util::{StreamExt, Stream};
+use async_channel::Receiver;
+use tokio::time::{Instant, MissedTickBehavior};
+use tokio_stream::wrappers::IntervalStream;
+
+use fluvio_types::SpuId;
+use fluvio_types::event::StickyEvent;
+use dataplane::ReplicaKey;
+use dataplane::batch::Batch;
+use dataplane::produce::{ProduceRequest, TopicProduceData, PartitionProduceData, ProduceResponse};
+
+use crate::FluvioError;
+use crate::spu::SpuPool;
+use crate::producer::{ProducerConfig, RecordPending};
+use crate::producer::partitioning::{Partitioner, SiphashRoundRobinPartitioner, PartitionerConfig};
+
+pub(crate) struct ProducerDispatcher {
+    /// A pool of connections to the SPUs.
+    pool: Arc<SpuPool>,
+    /// Configuration options controlling the behavior of the producer.
+    config: ProducerConfig,
+    /// The partitioning strategy to use
+    partitioner: Box<dyn Partitioner + Send + Sync>,
+    /// A buffer of records that have yet to be sent to the cluster.
+    buffer: BTreeMap<String, Vec<RecordPending>>, // (Topic, [Records])
+    /// A stream of events that the dispatcher needs to react to.
+    ///
+    /// A `NewRecord` event means that the [`Producer`]'s caller has called
+    /// `send` to queue a new record, and we need to buffer it until we're ready
+    /// to send.
+    ///
+    /// A `Timeout` event means we need to flush the buffer if there is anything in it.
+    events: Pin<Box<dyn Stream<Item = Event>>>,
+    /// A shutdown event used to determine when to quit.
+    ///
+    /// This will be triggered by the [`Producer`] that starts this dispatcher.
+    shutdown: Arc<StickyEvent>,
+}
+
+enum Event {
+    NewRecord((String, RecordPending<()>)),
+    Timeout(Instant),
+}
+
+impl ProducerDispatcher {
```

Comment:
follow dispatcher pattern in rest of the code.
and use similar flush logic in the SPU: https://github.com/infinyon/fluvio/blob/master/crates/fluvio-storage/src/mut_records.rs

Reply:
I've updated the dispatcher since this comment, it now respects `batch_size` and `batch_duration` when flushing